ls1intum/Artemis
# `General`: Enable score export with points per grading criterion into a CSV file

### Checklist
#### General


#### Server
#### Client

### Motivation and Context
We manually assess exercises according to a few grading criteria (e.g. functionality, understandability). The points received within a criterion are added up to yield a preliminary grade. Those are combined externally into an overall grade over multiple exercises. To make this possible, the previously used Praktomat supported the export of points values per criterion. Now Artemis is extended in that regard.

The overview of points might not only useful for this exact purpose but depending on the usage of grading criteria might give an insight which parts of the exercise were difficult for students, even if the overall grading does not directly depend on the points per criterion.

### Description
This PR extends the export scores into CSV functionality on the exercise score page. The summation of points per grading criterion is performed server-side.
For each criterion a new column with the criterion name is added to the CSV. For exercises that do nor use grading criteria the CSV only receives a new column ‘points’ in addition to the previously available ‘score’.

The CSV creation process in the client code has been simplified significantly. The tests now also check the format of the CSV, not only that one has been created.

### Steps for Testing
Prerequisites:
- An exercise with structured grading criteria and manual feedbacks.
- An exercise either without manual feedback or one without structured grading criteria.
- You can reuse existing exercises that meet those criteria and skip the parts of the steps for testing that create student submissions and manual feedbacks.

Steps for Testing:
1. As student, submit something to the exercise.
2. Log in as tutor/instructor after the exercise due date and manually assess the submissions.
   - When assessing, use some structured grading criteria instructions when creating the feedbacks.
   - You can also add feedbacks without using an instruction.
   - Keep track of the sum of points you set in the feedbacks per grading criterion.
3. Go to the scores page of the exercise.
4. Use the ‘Export Results’ button above the table to export the scores to a CSV file.
5. The resulting CSV should contain for each student:
   - Name
   - Login
   - Score
   - Total Points achieved in the exercise (should be the same as seen in the scores table)
   - For each grading criterion: a column with the name of the criterion and the sum of points the student achieved with feedbacks in that category
     (If the exercise has no grading criteria, then only the total points column should exist.)
   - If it is a programming exercise: a link to the student repository
6. When logged in as tutor, the link to the scores page should not be visible in the Course Management → Exercises view for any exercise type.

### Review Progress





#### Code Review
#### Manual Tests
(only test code has changed since the last manual tests)

### Test Coverage
| File | Branch | Line |
| --- | --- | --- |
| `ResultRepository.java` | 83% | 91% |
| `ResultResource.java` | 69% | 93% |
| `exercise-scores-export-button.component.ts` | 87.5% | 96.7% |
| `result.service.ts` (had no tests before, new code fully covered) | 61.3% | 62.8% |

### Screenshots
Instructor view of an exercise in the exercise list for a course: 
![instructor_view](https://user-images.githubusercontent.com/64250573/142601278-f57d5c85-9e9a-4e6d-8859-bc9ee35254e0.png)
Tutor view of an exercise in the exercise list for a course: 
![tutor_view](https://user-images.githubusercontent.com/64250573/142601279-731215d9-ed66-403f-8674-5e2f4367989b.png)
